### ISSUE TITLE
Unsupported target for indexed assignment for TypedDict union

### ISSUE THREAD

**Issue by dfroger** [python/mypy#10631]
```python
from typing_extensions import TypedDict
from typing import Union


class Foo1(TypedDict):
    z: str
    a: int


class Foo2(TypedDict):
    z: str
    b: int


def func(foo: Union[Foo1, Foo2]):
    foo["z"] = "z"
```

The error `Unsupported target for indexed assignment ("Union[Foo1, Foo2]")` is reported on line ` foo["z"] = "z"`. I would expect no error.

Tested with `mypy 0.902` and `typing-extensions 3.10.0.0`, no option given to `mypy`.

**Comment by dfroger** [comment]
it's working with `dataclasses`: mypy does not report error (and correctly reports error if the key do not exists).

Generally speaking, is static typing better supported for `dataclasses` than `TypedDict`?

```python
from dataclasses import dataclass
from typing import Union

@dataclass
class Foo1:
    z: str
    a: int


@dataclass
class Foo2:
    z: str
    b: int


def func(foo: Union[Foo1, Foo2]):
    foo.z = "z" 
```

**Comment by sobolevn** [comment]
Yes, looks like a bug a to me. Will provide a fix soon! Thanks! 👍

**Comment by sobolevn** [comment]
Yeap, this happens because of this line:

https://github.com/python/mypy/blob/209a7193feb4bbfa38d09232b0a5a916e9d2e605/mypy/checker.py#L3239-L3240

It works for `Foo1` and does not work for `Union[Foo1, ...]`

**Comment by sobolevn** [comment]
At the moment I am trying to route all `TypedDict` member access through `checkmember`. We use `getattr(typeddict_type, '__delitem__')` and `getattr(typeddict_type, '__setitem__')`. Not sure about `__getitem__`.

This does not go really well, since we also have a lot of plugin logic inside. I hope to finish this today.
